`AssertionError` in `floats(max_value=0, exclude_max=True)` when "denormals-are-zero" processor flag is set

From the normal Python shell and ipython, this works:

```python
$ ipython
Python 3.8.0 (default, Feb 25 2021, 22:10:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.27.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from hypothesis import strategies as st

In [2]: st.floats(max_value=0, exclude_max=True).example()
Out[2]: -5e-324

```

But from `./manage.py shell` and while actually running Django normally, it fails:

```python
$ ./manage.py shell
Python 3.8.0 (default, Feb 25 2021, 22:10:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.27.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from hypothesis import strategies as st

In [2]: st.floats(max_value=0, exclude_max=True).example()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-324381a1006a> in <module>
----> 1 st.floats(max_value=0, exclude_max=True).example()

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in example(self)
    330 
    331         examples: List[Ex] = []
--> 332         example_generating_inner_function()
    333         return random_choice(examples)
    334 

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in example_generating_inner_function()
    318         # tracebacks, and we want users to know that they can ignore it.
    319         @given(self)
--> 320         @settings(
    321             database=None,
    322             max_examples=10,

    [... skipping hidden 1 frame]

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/core.py in process_arguments_to_given(wrapped_test, arguments, kwargs, given_kwargs, argspec)
    447         search_strategy = WithRunner(search_strategy, selfy)
    448 
--> 449     search_strategy.validate()
    450 
    451     return arguments, kwargs, test_runner, search_strategy

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in validate(self)
    399         try:
    400             self.validate_called = True
--> 401             self.do_validate()
    402             self.is_empty
    403             self.has_reusable_values

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py in do_validate(self)
     44     def do_validate(self):
     45         for s in self.element_strategies:
---> 46             s.validate()
     47 
     48     def calc_label(self):

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in validate(self)
    399         try:
    400             self.validate_called = True
--> 401             self.do_validate()
    402             self.is_empty
    403             self.has_reusable_values

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in do_validate(self)
    799 
    800     def do_validate(self):
--> 801         self.mapped_strategy.validate()
    802 
    803     def pack(self, x):

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in validate(self)
    399         try:
    400             self.validate_called = True
--> 401             self.do_validate()
    402             self.is_empty
    403             self.has_reusable_values

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py in do_validate(self)
    133         w = self.wrapped_strategy
    134         assert isinstance(w, SearchStrategy), f"{self!r} returned non-strategy {w!r}"
--> 135         w.validate()
    136 
    137     def __repr__(self):

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in validate(self)
    399         try:
    400             self.validate_called = True
--> 401             self.do_validate()
    402             self.is_empty
    403             self.has_reusable_values

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in do_validate(self)
    799 
    800     def do_validate(self):
--> 801         self.mapped_strategy.validate()
    802 
    803     def pack(self, x):

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in validate(self)
    399         try:
    400             self.validate_called = True
--> 401             self.do_validate()
    402             self.is_empty
    403             self.has_reusable_values

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/collections.py in do_validate(self)
     44     def do_validate(self):
     45         for s in self.element_strategies:
---> 46             s.validate()
     47 
     48     def calc_label(self):

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/strategies.py in validate(self)
    399         try:
    400             self.validate_called = True
--> 401             self.do_validate()
    402             self.is_empty
    403             self.has_reusable_values

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py in do_validate(self)
    131 
    132     def do_validate(self):
--> 133         w = self.wrapped_strategy
    134         assert isinstance(w, SearchStrategy), f"{self!r} returned non-strategy {w!r}"
    135         w.validate()

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/lazy.py in wrapped_strategy(self)
    110             }
    111 
--> 112             base = self.function(*self.__args, **self.__kwargs)
    113             if unwrapped_args == self.__args and unwrapped_kwargs == self.__kwargs:
    114                 self.__wrapped_strategy = base

~/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/strategies/_internal/numbers.py in floats(min_value, max_value, allow_nan, allow_infinity, width, exclude_min, exclude_max)
    372             assert is_negative(max_value) and not is_negative(max_arg)
    373             max_value = next_down(max_value, width)
--> 374         assert max_value < max_arg  # type: ignore
    375 
    376     if min_value == -math.inf:

AssertionError: 

```

I'm definitely using the same installation of hypothesis from the same virtualenv:

```python
$ ipython
Python 3.8.0 (default, Feb 25 2021, 22:10:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.27.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from hypothesis import version

In [2]: version
Out[2]: <module 'hypothesis.version' from '/home/flyte/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/version.py'>

In [3]: version.__version__
Out[3]: '6.19.0'
```

```python
$ ./manage.py shell
Python 3.8.0 (default, Feb 25 2021, 22:10:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.27.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from hypothesis import version

In [2]: version
Out[2]: <module 'hypothesis.version' from '/home/flyte/.local/share/virtualenvs/myproject-tDErfRdz/lib/python3.8/site-packages/hypothesis/version.py'>

In [3]: version.__version__
Out[3]: '6.19.0'
```

I'm using Django==3.1.13

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,11 @@
+RELEASE_TYPE: minor
+
+This release makes :func:`~hypothesis.strategies.floats` error *consistently* when
+your floating-point hardware has been configured to violate IEEE-754 for
+:wikipedia:`subnormal numbers <Subnormal_number>`, instead of
+only when an internal assertion was tripped (:issue:`3092`).
+
+If this happens to you, passing ``allow_subnormal=False`` will suppress the explicit
+error.  However, we strongly recommend fixing the root cause by disabling global-effect
+unsafe-math compiler options instead, or at least consulting e.g. Simon Byrne's
+`Beware of fast-math <https://simonbyrne.github.io/notes/fastmath/>`__ explainer first.

--- a/hypothesis-python/src/hypothesis/strategies/_internal/numbers.py
+++ b/hypothesis-python/src/hypothesis/strategies/_internal/numbers.py
@@ -342,6 +342,22 @@ def floats(
     check_valid_bound(min_value, "min_value")
     check_valid_bound(max_value, "max_value")
 
+    if allow_subnormal and next_up(0.0, width=width) == 0:  # pragma: no cover
+        # Not worth having separate CI envs and dependencies just to cover this branch;
+        # discussion in https://github.com/HypothesisWorks/hypothesis/issues/3092
+        #
+        # Erroring out here ensures that the database contents are interpreted
+        # consistently - which matters for such a foundational strategy, even if it's
+        # not always true for all user-composed strategies further up the stack.
+        raise FloatingPointError(
+            f"Got allow_subnormal={allow_subnormal!r}, but we can't represent "
+            f"subnormal floats right now, in violation of the IEEE-754 floating-point "
+            f"specification.  This is usually because something was compiled with "
+            f"-ffast-math or a similar option, which sets global processor state.  "
+            f"See https://simonbyrne.github.io/notes/fastmath/ for a more detailed "
+            f"writeup - and good luck!"
+        )
+
     min_arg, max_arg = min_value, max_value
     if min_value is not None:
         min_value = float_of(min_value, width)

